File: src/scripts/download-github-images.py
#!/usr/bin/env python3
"""
Script alternativo in Python per scaricare le immagini da GitHub
Utile se Node.js non funziona correttamente
"""

import os
import sys
import json
import urllib.request
import urllib.parse
from pathlib import Path

# Configurazione
GITHUB_REPO = "salvatori780-bit/imagesportfoliooo"
GITHUB_BRANCH = "main"
GITHUB_PATH = "prog. figma"
OUTPUT_DIR = Path(__file__).parent.parent / "public" / "images" / "projects"

# URLs
GITHUB_API_URL = f"https://api.github.com/repos/{GITHUB_REPO}/contents/{urllib.parse.quote(GITHUB_PATH)}?ref={GITHUB_BRANCH}"
GITHUB_RAW_BASE = f"https://raw.githubusercontent.com/{GITHUB_REPO}/{GITHUB_BRANCH}/{urllib.parse.quote(GITHUB_PATH)}"

def print_header():
    print("🚀 GitHub Images Downloader (Python)")
    print("━" * 60)
    print()
    print(f"📦 Repository: {GITHUB_REPO}")
    print(f"🌿 Branch: {GITHUB_BRANCH}")
    print(f"📁 Path: {GITHUB_PATH}")
    print(f"💾 Output: {OUTPUT_DIR}")
    print()

def create_output_dir():
    """Crea directory output se non esiste"""
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    print(f"✅ Directory ready: {OUTPUT_DIR}")
    print()

def get_file_list():
    """Ottieni lista file da GitHub API"""
    try:
        print("📡 Fetching file list from GitHub...")
        
        req = urllib.request.Request(GITHUB_API_URL)
        req.add_header('User-Agent', 'Python Script')
        
        with urllib.request.urlopen(req) as response:
            data = response.read()
            files = json.loads(data)
            return files
            
    except Exception as e:
        print(f"❌ Error fetching file list: {e}")
        return []

def download_file(url, output_path):
    """Scarica un file da URL"""
    try:
        urllib.request.urlretrieve(url, output_path)
        return True
    except Exception as e:
        print(f"   ❌ Error: {e}")
        return False

def main():
    print_header()
    create_output_dir()
    
    # Ottieni lista file
    files = get_file_list()
    
    if not files:
        print("❌ Failed to fetch file list")
        print()
        print("💡 Troubleshooting:")
        print("   1. Check internet connection")
        print("   2. Verify repository is public")
        print("   3. Check the path is correct")
        sys.exit(1)
    
    # Filtra solo PNG
    image_files = [f for f in files if f.get('type') == 'file' and f['name'].lower().endswith('.png')]
    
    print(f"✅ Found {len(image_files)} PNG files")
    print()
    
    if not image_files:
        print("⚠️  No PNG files found")
        sys.exit(0)
    
    # Scarica ogni immagine
    downloaded = 0
    failed = 0
    
    for file_info in image_files:
        filename = file_info['name']
        download_url = file_info['download_url']
        output_path = OUTPUT_DIR / filename
        
        print(f"⬇️  Downloading: {filename}...")
        
        if download_file(download_url, output_path):
            downloaded += 1
            # Ottieni dimensione file
            size_kb = output_path.stat().st_size / 1024
            print(f"   ✅ Saved ({size_kb:.1f} KB)")
        else:
            failed += 1
    
    # Summary
    print()
    print("━" * 60)
    print("📊 Download Summary")
    print("━" * 60)
    print(f"✅ Downloaded: {downloaded} files")
    print(f"❌ Failed: {failed} files")
    print(f"📁 Location: {OUTPUT_DIR}")
    print()
    
    if downloaded > 0:
        print("🎉 Images downloaded successfully!")
        print()
        print("📝 Next steps:")
        print("   1. Review downloaded images")
        print("   2. Run: npm run update-imports")
        print("   3. Test: npm run dev")
        print()
    
    sys.exit(0 if failed == 0 else 1)

if __name__ == "__main__":
    main()

File: src/scripts/download-github-images.js
#!/usr/bin/env node

/**
 * Script per scaricare le immagini dal repository GitHub
 * e sostituirle nel progetto
 * 
 * Repo: https://github.com/salvatori780-bit/imagesportfoliooo/tree/main/prog.%20figma
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

// Configurazione
const GITHUB_REPO = 'salvatori780-bit/imagesportfoliooo';
const GITHUB_BRANCH = 'main';
const GITHUB_PATH = 'prog. figma'; // Nota lo spazio!
const OUTPUT_DIR = path.join(__dirname, '..', 'public', 'images', 'projects');

// API GitHub per ottenere la lista dei file
const GITHUB_API_URL = `https://api.github.com/repos/${GITHUB_REPO}/contents/${encodeURIComponent(GITHUB_PATH)}?ref=${GITHUB_BRANCH}`;

// URL raw per scaricare i file
const GITHUB_RAW_BASE = `https://raw.githubusercontent.com/${GITHUB_REPO}/${GITHUB_BRANCH}/${encodeURIComponent(GITHUB_PATH)}`;

console.log('🚀 GitHub Images Downloader');
console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
console.log('');
console.log(`📦 Repository: ${GITHUB_REPO}`);
console.log(`🌿 Branch: ${GITHUB_BRANCH}`);
console.log(`📁 Path: ${GITHUB_PATH}`);
console.log(`💾 Output: ${OUTPUT_DIR}`);
console.log('');

// Crea directory output se non esiste
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  console.log(`✅ Created directory: ${OUTPUT_DIR}`);
}

/**
 * Scarica un file da URL
 */
function downloadFile(url, outputPath) {
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(outputPath);
    
    https.get(url, (response) => {
      if (response.statusCode === 200) {
        response.pipe(file);
        file.on('finish', () => {
          file.close();
          resolve();
        });
      } else if (response.statusCode === 302 || response.statusCode === 301) {
        // Segui redirect
        downloadFile(response.headers.location, outputPath)
          .then(resolve)
          .catch(reject);
      } else {
        fs.unlink(outputPath, () => {});
        reject(new Error(`Failed to download: ${response.statusCode}`));
      }
    }).on('error', (err) => {
      fs.unlink(outputPath, () => {});
      reject(err);
    });
  });
}

/**
 * Ottieni lista file da GitHub
 */
function getFileList() {
  return new Promise((resolve, reject) => {
    https.get(GITHUB_API_URL, {
      headers: {
        'User-Agent': 'Node.js Script'
      }
    }, (response) => {
      let data = '';
      
      response.on('data', (chunk) => {
        data += chunk;
      });
      
      response.on('end', () => {
        if (response.statusCode === 200) {
          try {
            const files = JSON.parse(data);
            resolve(files);
          } catch (err) {
            reject(new Error('Failed to parse GitHub API response'));
          }
        } else {
          reject(new Error(`GitHub API error: ${response.statusCode}`));
        }
      });
    }).on('error', reject);
  });
}

/**
 * Main function
 */
async function main() {
  try {
    console.log('📡 Fetching file list from GitHub...');
    const files = await getFileList();
    
    // Filtra solo le immagini PNG
    const imageFiles = files.filter(file => 
      file.type === 'file' && 
      file.name.toLowerCase().endsWith('.png')
    );
    
    console.log(`✅ Found ${imageFiles.length} PNG files`);
    console.log('');
    
    if (imageFiles.length === 0) {
      console.log('⚠️  No PNG files found in the repository');
      return;
    }
    
    // Scarica ogni immagine
    let downloaded = 0;
    let failed = 0;
    
    for (const file of imageFiles) {
      const outputPath = path.join(OUTPUT_DIR, file.name);
      
      try {
        console.log(`⬇️  Downloading: ${file.name}...`);
        await downloadFile(file.download_url, outputPath);
        downloaded++;
        console.log(`   ✅ Saved to: ${outputPath}`);
      } catch (err) {
        failed++;
        console.log(`   ❌ Failed: ${err.message}`);
      }
    }
    
    console.log('');
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log('📊 Download Summary');
    console.log('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
    console.log(`✅ Downloaded: ${downloaded} files`);
    console.log(`❌ Failed: ${failed} files`);
    console.log(`📁 Location: ${OUTPUT_DIR}`);
    console.log('');
    
    if (downloaded > 0) {
      console.log('🎉 Images downloaded successfully!');
      console.log('');
      console.log('📝 Next steps:');
      console.log('   1. Review downloaded images in: ' + OUTPUT_DIR);
      console.log('   2. Run: npm run update-image-imports');
      console.log('   3. Test the application');
    }
    
  } catch (err) {
    console.error('');
    console.error('❌ Error:', err.message);
    console.error('');
    console.error('💡 Troubleshooting:');
    console.error('   1. Check your internet connection');
    console.error('   2. Verify the GitHub repository is accessible');
    console.error('   3. Check the repository path: ' + GITHUB_PATH);
    process.exit(1);
  }
}

// Run
main();
